src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -84,14 +84,132 @@ private static IEnumerable<TextEdit> FormatRazor(FormattingContext context, Razo
                 TryFormatSingleLineDirective(context, edits, source, node);
+                TryFormatBlocks(context, edits, source, node);
             }
 
             return edits;
         }
 
-        private static void TryFormatCSharpCodeBlock(FormattingContext context, List<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        private static bool TryFormatBlocks(FormattingContext context, IList<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        {
+            return TryFormatFunctionsBlock(context, edits, source, node) ||
+                TryFormatCSharpExplicitTransition(context, edits, source, node) ||
+                TryFormatHtmlInCSharp(context, edits, source, node) ||
+                TryFormatComplexCSharpBlock(context, edits, source, node);
+        }
+
+        private static bool TryFormatFunctionsBlock(FormattingContext context, IList<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        {
+            // @functions
+            // {
+            // }
+            if (node is CSharpCodeBlockSyntax directiveCode &&
+                directiveCode.Children.Count == 1 && directiveCode.Children.First() is RazorDirectiveSyntax directive &&
+                directive.Body is RazorDirectiveBodySyntax directiveBody &&
+                directiveBody.Keyword.GetContent().Equals("functions"))
+            {
+                var cSharpCode = directiveBody.CSharpCode;
+                var openBrace = cSharpCode.Children.FirstOrDefault(c => c.Kind == SyntaxKind.RazorMetaCode && c.GetContent().Equals("{", StringComparison.Ordinal));
+                var closeBrace = cSharpCode.Children.LastOrDefault(c => c.Kind == SyntaxKind.RazorMetaCode && c.GetContent().Equals("}", StringComparison.Ordinal));
+                var code = cSharpCode.Children.First(c => c.Kind == SyntaxKind.CSharpCodeBlock) as CSharpCodeBlockSyntax;
+
+                if (openBrace is null || closeBrace is null)
+                {
+                    // Don't trust ourselves in an incomplete scenario.
+                    return false;
+                }
+
+                var openBraceNode = openBrace;
+                var codeNode = code!;
+                var closeBraceNode = closeBrace;
+
+                return FormatBlock(context, source, openBraceNode, codeNode, closeBraceNode, edits, additionalIndentation: false);
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private static bool TryFormatCSharpExplicitTransition(FormattingContext context, IList<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        {
+            // We're looking for a code block like this:
+            //
+            // @{
+            //     var x = 1;
+            // }
+            if (node is CSharpCodeBlockSyntax expliciteCode &&
+                expliciteCode.Children.First() is CSharpStatementSyntax statement &&
```

Comment:
Is `.First()` dangerous? Or are we always guaranteed to have a child?

Reply:
First is probably dangerous, will change.